Onboarding: Intermediate lectures (+ runnable Aiken and Mesh examples) (#1951)
cardano-foundation/developer-portal: docs/developers/onboarding/lectures/intermediate/4-datum-and-redeemer.md

---
title: "Datum & redeemer"
sidebar_label: "Datum & redeemer"
description: "The two pieces of data a validator works with: the datum locked with a UTxO, and the redeemer the spender provides."
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import CodeBlock from "@theme/CodeBlock";
import extractRegion from "@site/src/utils/extractRegion";
import VaultSimple from "!!raw-loader!@site/examples/onboarding/lectures/intermediate/vault/on-chain/aiken/validators/vault_simple.ak";

# Datum & redeemer

[Last lecture](/docs/developers/onboarding/lectures/intermediate/what-is-a-validator) said a validator is a function of **datum**, **redeemer**, and **context**. The validator you wrote ignores all three. This lecture is about the first two, which are how you give information to a contract. They are also the part newcomers most often confuse, so they are worth explaining carefully.

> - The **datum** is information **attached to the locked UTxO** when you lock it. These are the _terms_. Think of it as a note that says "this is locked under these conditions". It is fixed the moment the funds are locked and never changes.

Reviewer: ```suggestion
- The **datum** is information attached to the UTxO when you **lock it**. You can put anything in there. Addresses, numbers, information, anything! The important part is that it is fixed the moment the funds are locked and never changes.
```
I think referring to the datums as "terms" or "conditions" could be problematic because those terms could be interpreted as "rules," which you established are the validator, not the datum.

> - The **redeemer** is what the **spender provides** when they try to unlock. It is their _choice_ for this attempt, and it is supplied fresh in the spending transaction.

Reviewer: ```suggestion
- The **redeemer** is what the **spender provides** when they try to unlock. It is their _choice_ for this attempt (transaction), and they supply it fresh in the spending transaction.
```

> - The **context** is the rest of the transaction: its inputs, outputs, signatures, and the validity window from [Time on Cardano](/docs/developers/onboarding/lectures/beginner/time-on-cardano). The validator can read all of it. There is enough of it to fill [the next lecture](/docs/developers/onboarding/lectures/intermediate/transaction-context) on its own, so this one is about the first two.

Reviewer: ```suggestion
- The **context** is the rest of the transaction: its inputs, outputs, signatures, the validity window from [Time on Cardano](/docs/developers/onboarding/lectures/beginner/time-on-cardano), and more. The validator can read all of it. There is enough info to fill [the next lecture](/docs/developers/onboarding/lectures/intermediate/transaction-context) on its own, so this one is about the first two.
```

    Note over You,Vault: Transaction 1, locking
    You->>Net: sign + submit a payment to the script address,<br/>with the datum attached
    Net->>Vault: an ordinary payment, accepted. The 5 ADA sits here
    Note over Net: the validator does not run:<br/>nothing is being unlocked yet
    Note over Vault: 5 ADA + datum (the terms): owner = your key hash

    Note over You,Vault: later. Anyone may try to spend that UTxO

    Note over You,Vault: Transaction 2, unlocking
    You->>Net: sign + submit a spend of that UTxO,<br/>with the redeemer attached
    Net->>Net: run the validator on three things:<br/>the datum (read off the UTxO), the redeemer (from this transaction),<br/>and the context (this transaction itself)
    Net-->>You: True, or False
    Note over Net,You: True → the 5 ADA moves where transaction 2 says<br/>False → transaction 2 is rejected, the UTxO stays put
```

Two transactions, and only the second one is judged. Everything the **datum** says was settled in
transaction 1, by whoever locked the funds, and it cannot be changed now. Everything the **redeemer**
says is what the spender brings today, in transaction 2. The validator's whole job is to check the

> second against the first, in the situation the context describes.

Reviewer: This part is confusing; I think we should rethink it.

Plus, it has some issues:
- " "I am here to collect the bag." That request is the redeemer. ", that's the spending transaction, not the redeemer.
- Diagram last step: It's not clear that you're releasing the datum created in the script address. Both cases come back from "network" to "Your app".
- It's not clear if this is just an example about datums or if you're explaining the first protocol.

Author: Diagram updated
